core/src/peer_record.rs

```diff
@@ -129,7 +126,7 @@ pub enum FromEnvelopeError {
     /// Failed to extract the payload from the envelope.
     BadPayload(signed_envelope::ReadPayloadError),
     /// Failed to decode the provided bytes as a [`PeerRecord`].
-    InvalidPeerRecord(prost::DecodeError),
+    InvalidPeerRecord(protobuf::Error),
```

Comment:
Is this part of the public API? If yes, then this is actually a breaking change which is unfortunate :/

Reply:
Oh I missed that. Yes, it is part of the public API :disappointed:.

Reply:
Damn okay. Well don't worry about it here for now, I'll send a PR that wraps the error here. That is still a breaking change but we can ship that separately!

Reply:
So my idea here is to make a dedicated error like this:

```rust
#[derive(thiserror::Error, Debug)]
#[error(transparent)]
pub struct DecodeError(protobuf::Error)
```

This preserves the source chain but hides the concrete error type from the public API.

It'd be nice to do this in a separate PR and then put this feature on top of that. Let me know if you'd be happy to do that otherwise I'll try to get to that next week!

Reply:
> So my idea here is to make a dedicated error like this:
> 
> ```rust
> #[derive(thiserror::Error, Debug)]
> #[error(transparent)]
> pub struct DecodeError(protobuf::Error)
> ```
> 
> This preserves the source chain but hides the concrete error type from the public API.

That sounds good! 

> It'd be nice to do this in a separate PR and then put this feature on top of that. Let me know if you'd be happy to do that otherwise I'll try to get to that next week!

If you could please do that that would be great! I'll rebase once that is merged.

Reply:
I starting working on this here: https://github.com/libp2p/rust-libp2p/pull/3058